pkg/object/graph/depends.go

```diff
@@ -83,20 +83,25 @@ func ReverseSortObjs(objs object.UnstructuredSet) ([]object.UnstructuredSet, err
 // addApplyTimeMutationEdges updates the graph with edges from objects
 // with an explicit "apply-time-mutation" annotation.
 func addApplyTimeMutationEdges(g *Graph, objs object.UnstructuredSet) {
+	ids := object.UnstructuredsToObjMetasOrDie(objs)
 	for _, obj := range objs {
 		id := object.UnstructuredToObjMetaOrDie(obj)
 		klog.V(3).Infof("adding vertex: %s", id)
 		g.AddVertex(id)
 		if mutation.HasAnnotation(obj) {
 			subs, err := mutation.ReadAnnotation(obj)
 			if err != nil {
-				// TODO: fail task if parse errors?
+				// Log & skip - Should have been caught by validation
 				klog.V(3).Infof("failed to add edges from: %s: %s", id, err)
-				return
+				continue
 			}
 			for _, sub := range subs {
-				// TODO: fail task if it's not in the inventory?
 				dep := sub.SourceRef.ObjMetadata()
+				if !ids.Contains(dep) {
```

Comment:
Could we handle the validation here where it is needed, rather than doing all validation early? I realize doing it all early is beneficial for detecting all errors early and then report them all in an early exit, but it seems like some validation errors will only be caught later. Circular dependencies would be one of them.

Reply:
I think I can make SortObjs return a list of validation errors, without exiting on each one. Then we can move the SortObjs call from the solver to the applier/destroyer and send the ValidationEvents from there.